src/main/java/com/aws/greengrass/lifecyclemanager/KernelLifecycle.java

```diff
@@ -366,6 +328,68 @@ void initConfigAndTlog() {
         }
     }
 
+    /*
+     * Check if last tlog truncation was interrupted and undo its effect
+     *
+     * @param transactionLogPath path to config.tlog
+     * @return true if last tlog truncation was complete or if we are able to undo its effect;
+     *         false only if there was an IO error while undoing its effect (renaming the old tlog file)
+     */
+    private boolean handleIncompleteTlogTruncation(Path transactionLogPath) {
+        Path oldTlogPath = ConfigurationWriter.getOldTlogPath(transactionLogPath);
+        // At the beginning of tlog truncation, the original config.tlog file is moved to config.tlog.old
+        // If .old file exists, then the last truncation was incomplete, so we need to undo its effect by moving it
+        // back to the original location.
+        if (Files.exists(oldTlogPath)) {
```

Comment:
My understanding is that, if the .old file exists, then the truncation must have failed.

In this case I trust the .old file more because it's directly moved from last valid config. I don't trust the config.tlog as much even if it exists because it must have been created sometime during truncation and more likely to be incomplete.